### ISSUE TITLE
Add support for PostgreSQL's DELETE .. USING syntax

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#4245]
This is just a fancy way of expressing a semi-join:

``` sql
DELETE FROM films USING producers
  WHERE producer_id = producers.id AND producers.name = 'foo';
```

is the same as:

``` sql
DELETE FROM films
  WHERE producer_id IN (SELECT id FROM producers WHERE name = 'foo');
```

... and it can be emulated as such, in other databases.

See: http://www.postgresql.org/docs/9.4/static/sql-delete.html

**Comment by icl7126** [comment]
It's not the same thing, there may be performance differences. Original documentation is mentioning it:
`In some cases the join style is easier to write or faster to execute than the sub-select style.`
I agree this is a low level priority but it would be really nice to have.

**Comment by lukaseder** [comment]
Thanks for the feedback. We might re-consider implementing this, although, no promises. It seems like a rather non-idiomatic way to write a `JOIN`. Other databases do support actual `JOIN` expressions in `DELETE`

**Comment by tomdcc** [comment]
Another vote for this syntax, I just hit a need to do this.

**Comment by tomdcc** [comment]
For the record, these all produce different plans for my db:

```sql
DELETE FROM target_table
USING other_table
WHERE target_other_id = other_id
AND other_thing LIKE 'some other condition%';
```

```sql
DELETE FROM target_table
WHERE NOT EXISTS (
  SELECT 1 FROM other_table
  WHERE target_other_id = other_id
  AND NOT other_thing LIKE 'some other condition%'
);
```

```sql
DELETE FROM target_table
WHERE target_other_id IN (
  SELECT other_id
  FROM other_table
  WHERE other_thing LIKE 'some other condition%'
);
```

So there are performance reasons to want to do this.

**Comment by lukaseder** [comment]
I get three times the same explained execution plan for these three queries on the [Sakila database](https://www.jooq.org/sakila) on PostgreSQL 9.6.5:

```sql
DELETE FROM film_actor
USING actor
WHERE film_actor.actor_id = actor.actor_id
AND last_name LIKE 'ABC%';

DELETE FROM film_actor
WHERE EXISTS (
  SELECT 1
  FROM actor
  WHERE film_actor.actor_id = actor.actor_id
  AND last_name LIKE 'ABC%'
);

DELETE FROM film_actor
WHERE actor_id IN (
  SELECT actor_id
  FROM actor
  WHERE last_name LIKE 'ABC%'
);
```

Note, your `NOT EXISTS` version is not equivalent to the others.

So, I'm still quite reluctant to introduce this rather funky syntax.

**Comment by fkalinski** [comment]
On database I'm working with (PostgreSQL 9.6) the following operations have different query plans
```
delete from d_item_daily_history idh
using (
    select id,
           row_number() over (partition by item_id order by content_date desc) rn
      from d_item_daily_history
     where content_date < $2
       and warehouse_group_id = $1
) h
where h.id = idh.id 
  and h.rn > 1; 

Delete on d_item_daily_history idh  (cost=589742.15..965795.83 rows=47383 width=38) (actual time=2929.608..2929.608 rows=0 loops=1)
  ->  Nested Loop  (cost=589742.15..965795.83 rows=47383 width=38) (actual time=389.161..1901.114 rows=416712 loops=1)
        ->  Subquery Scan on h  (cost=589741.59..594361.40 rows=47383 width=40) (actual time=389.138..850.382 rows=416712 loops=1)
              Filter: (h.r > 1)
              Rows Removed by Filter: 138904
              ->  WindowAgg  (cost=589741.59..592584.55 rows=142148 width=28) (actual time=389.111..707.079 rows=555616 loops=1)
                    ->  Sort  (cost=589741.59..590096.96 rows=142148 width=20) (actual time=389.103..458.886 rows=555616 loops=1)
                          Sort Key: idh_1.item_id, idh_1.content_date DESC
                          Sort Method: quicksort  Memory: 67984kB
                          ->  Bitmap Heap Scan on d_item_daily_history idh_1  (cost=12147.97..577575.83 rows=142148 width=20) (actual time=32.507..149.819 rows=555616 loops=1)
                                Recheck Cond: (warehouse_group_id = '2517'::bigint)
                                Filter: (content_date < '2020-01-08'::date)
                                Heap Blocks: exact=5781
                                ->  Bitmap Index Scan on idx_d_daily_i_hist_whg_id  (cost=0.00..12112.44 rows=564783 width=0) (actual time=31.623..31.623 rows=555616 loops=1)
                                      Index Cond: (warehouse_group_id = '2517'::bigint)
        ->  Index Scan using pk_d_daily_i_history on d_item_daily_history idh  (cost=0.56..7.83 rows=1 width=14) (actual time=0.002..0.002 rows=1 loops=416712)
              Index Cond: (id = h.id)
Execution time: 2933.342 ms
```
and
```
delete from d_item_daily_history 
where id in (
  select id from (
    select id, 
           row_number() over (partition by item_id order by content_date desc) rn
      from d_item_daily_history
     where content_date < $2
       and warehouse_group_id = $1
  ) t where t.rn > 1 
);

Delete on d_item_daily_history idh  (cost=594480.42..966388.11 rows=47383 width=38) (actual time=3357.655..3357.655 rows=0 loops=1)
  ->  Nested Loop  (cost=594480.42..966388.11 rows=47383 width=38) (actual time=923.100..2345.532 rows=416712 loops=1)
        ->  HashAggregate  (cost=594479.86..594953.69 rows=47383 width=40) (actual time=923.074..1101.913 rows=416712 loops=1)
              Group Key: t.id
              ->  Subquery Scan on t  (cost=589741.59..594361.40 rows=47383 width=40) (actual time=393.144..780.152 rows=416712 loops=1)
                    Filter: (t.r > 1)
                    Rows Removed by Filter: 138904
                    ->  WindowAgg  (cost=589741.59..592584.55 rows=142148 width=28) (actual time=393.116..668.445 rows=555616 loops=1)
                          ->  Sort  (cost=589741.59..590096.96 rows=142148 width=20) (actual time=393.108..469.746 rows=555616 loops=1)
                                Sort Key: d_item_daily_history.item_id, d_item_daily_history.content_date DESC
                                Sort Method: quicksort  Memory: 67984kB
                                ->  Bitmap Heap Scan on d_item_daily_history  (cost=12147.97..577575.83 rows=142148 width=20) (actual time=34.210..151.192 rows=555616 loops=1)
                                      Recheck Cond: (warehouse_group_id = '2517'::bigint)
                                      Filter: (content_date < '2020-01-08'::date)
                                      Heap Blocks: exact=5781
                                      ->  Bitmap Index Scan on idx_d_daily_i_hist_whg_id  (cost=0.00..12112.44 rows=564783 width=0) (actual time=33.325..33.325 rows=555616 loops=1)
                                            Index Cond: (warehouse_group_id = '2517'::bigint)
        ->  Index Scan using pk_d_daily_i_history on d_item_daily_history idh  (cost=0.56..7.83 rows=1 width=14) (actual time=0.003..0.003 rows=1 loops=416712)
              Index Cond: (id = t.id)
Execution time: 3366.326 ms
```

So it's not just a syntax sugar.

**Comment by lukaseder** [comment]
> So it's not just a syntax sugar.

Aah, the promise of 4GL in the hopes of perfect optimisers. It should not be necessary to manually rewrite provably equivalent SQL, but we still have to, in 2020. I've written a long blog post on quite a few non-cost-based optimisations that RDBMS *should* implement, but some (including PostgreSQL) are quite reluctant to do so:
https://blog.jooq.org/2017/09/28/10-cool-sql-optimisations-that-do-not-depend-on-the-cost-model/

In the case of PostgreSQL, the strategy is to not invest heavily in execution plan caching (such as DB2, Oracle, SQL Server), but re-plan queries much more frequently, in case of which SQL transformations become much more expensive, as they have to be re-done every time.

Alas. We'll reconsider adding this...

**Comment by lukaseder** [comment]
The emulation in other dialects will happen in a separate issue: #9810

**Comment by lukaseder** [comment]
The API is similar to that of `SelectFromStep`. Fun fact, `USING` is included in the projection of `RETURNING`:

```sql
create table t as select 1 a union select 2 a union select 3 a;

delete from t 
using (values(1)) u(b), (values(2)) v(c)
where t.a = u.b or t.a = v.c
returning *;
```

Yields

|a|b|c|
|-|-|-|
|1|1|2|
|2|1|2|

**Comment by lukaseder** [comment]
... this obviously makes it almost impossible to emulate `DELETE .. USING` in the presence of `RETURNING`.

**Comment by lukaseder** [comment]
OK, native PostgreSQL will be available starting with jOOQ 3.13 (released soon)
